File: PayPal/src/main/java/com/braintreepayments/api/PayPalPaymentAuthRequestParams.java
package com.braintreepayments.api;

public class PayPalPaymentAuthRequestParams {

    private String approvalUrl;
    private String clientMetadataId;
    private String pairingId;
    private final PayPalRequest payPalRequest;

    private String successUrl;

    private BrowserSwitchOptions browserSwitchOptions;

    PayPalPaymentAuthRequestParams(PayPalRequest payPalRequest) {
        this.payPalRequest = payPalRequest;
    }

    String getApprovalUrl() {
        return approvalUrl;
    }

    PayPalPaymentAuthRequestParams approvalUrl(String value) {
        approvalUrl = value;
        return this;
    }

    String getClientMetadataId() {
        return clientMetadataId;
    }

    PayPalPaymentAuthRequestParams clientMetadataId(String value) {
        clientMetadataId = value;
        return this;
    }

    String getIntent() {
        if (payPalRequest instanceof PayPalCheckoutRequest) {
            return ((PayPalCheckoutRequest) payPalRequest).getIntent();
        }
        return null;
    }

    String getUserAction() {
        if (payPalRequest instanceof PayPalCheckoutRequest) {
            return ((PayPalCheckoutRequest) payPalRequest).getUserAction();
        }
        return "";
    }

    boolean isBillingAgreement() {
        return payPalRequest instanceof PayPalVaultRequest;
    }

    String getMerchantAccountId() {
        return payPalRequest.getMerchantAccountId();
    }

    String getPairingId() {
        return pairingId;
    }

    PayPalPaymentAuthRequestParams pairingId(String value) {
        pairingId = value;
        return this;
    }

    String getSuccessUrl() {
        return successUrl;
    }

    PayPalPaymentAuthRequestParams successUrl(String value) {
        successUrl = value;
        return this;
    }

    BrowserSwitchOptions getBrowserSwitchOptions() {
        return browserSwitchOptions;
    }

    void setBrowserSwitchOptions(BrowserSwitchOptions browserSwitchOptions) {
        this.browserSwitchOptions = browserSwitchOptions;
    }
}

File: PayPal/src/main/java/com/braintreepayments/api/paypal/PayPalPaymentAuthRequestParams.kt
package com.braintreepayments.api.paypal

import androidx.annotation.RestrictTo
import com.braintreepayments.api.BrowserSwitchOptions

@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
data class PayPalPaymentAuthRequestParams @JvmOverloads internal constructor(
    private val payPalRequest: PayPalRequest,
    var browserSwitchOptions: BrowserSwitchOptions?,
    var approvalUrl: String? = null,
    val clientMetadataId: String? = null,
    val contextId: String? = null,
    val successUrl: String? = null,
) {

    @get:RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    val intent: PayPalPaymentIntent?
        get() = if (payPalRequest is PayPalCheckoutRequest) payPalRequest.intent else null

    @get:RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    val isVaultRequest: Boolean
        get() = payPalRequest is PayPalVaultRequest

    @get:RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    val merchantAccountId: String?
        get() = payPalRequest.merchantAccountId

    @get:RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    val fundingSource: String
        get() = payPalRequest.getFundingSource().value

    @get:RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    val shouldRequestBillingAgreement: Boolean
        get() = payPalRequest is PayPalCheckoutRequest && payPalRequest.shouldRequestBillingAgreement

    @get:RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    val recurringBillingPlanType: String?
        get() = payPalRequest.recurringBillingPlanType?.name
}
